src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_scan_info(channel)) {
+            SOL_WRN("Could not read scanning info of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        if (!read_channel_index_in_buffer(channel)) {
+            SOL_WRN("Could not read index in buffer of channel [%s] in device%d",
+                channel->name, device->device_id);
+            goto error;
+        }
+
+        channel->offset_in_buffer = -1;
+    }
+
+    sol_ptr_vector_append(&channel->device->channels, channel);
+
+    SOL_DBG("channel %s added. scale: %lf - offset: %d - storagebits: %d",
+        channel->name, channel->scale, channel->offset, channel->storagebits);
+
+    return channel;
+
+error:
+    iio_del_channel(channel);
+
+    return NULL;
+}
+
+bool
+sol_iio_read_channel_value(struct sol_iio_channel *channel, double *value)
+{
+    int len, raw_value;
+    char path[PATH_MAX + 1];
+    struct sol_iio_device *device = channel->device;
+
+    if (device->buffer_enabled) {
+        SOL_WRN("device%d is using buffer - use iio_read_buffer_channel_value() instead",
+            device->device_id);
+        return false;
+    }
+
+    len = snprintf(path, sizeof(path), CHANNEL_RAW_PATH, device->device_id,
+        channel->name);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    len = sol_util_read_file(path, "%d", &raw_value);
+    if (len < 0) {
+        SOL_WRN("Could not read channel [%s] in device%d", channel->name,
+            device->device_id);
+        return false;
+    }
+
+    *value = (raw_value + channel->offset) * channel->scale;
+    return true;
+}
+
+void
+sol_iio_set_reader_cb(struct sol_iio_device *device, sol_iio_reader_cb reader_cb, void *data)
+{
+    device->reader_cb = reader_cb;
+    device->reader_cb_data = data;
+}
+
+static int
+calc_channel_offset_in_buffer(struct sol_iio_channel *channel)
+{
+    struct sol_iio_channel *channel_itr;
+    int i, offset = 0;
+
+    SOL_PTR_VECTOR_FOREACH_IDX (&channel->device->channels, channel_itr, i) {
+        if (channel_itr->index < channel->index)
+            offset += channel->storagebits;
+    }
+
+    return offset;
+}
+
+bool
+sol_iio_read_buffer_channel_value(struct sol_iio_channel *channel, struct sol_iio_buffer *buffer, double *value)
+{
+    uint64_t data = 0;
+    int64_t s_data = -1;
+    int i, j, offset_bytes, storage_bytes;
+    bool negative = false;
+
+    /* Leve this calc to the first read, as now we know how much channels are enabled */
+    if (channel->offset_in_buffer == -1)
+        channel->offset_in_buffer = calc_channel_offset_in_buffer(channel);
+
+    if (channel->offset_in_buffer + channel->storagebits > (int)buffer->size * 8) {
```

Comment:
i guess you need this cast `(int)` because of signed x unsigned. Make `offset_in_buffer` and `channel->storagebits` all unsigned and you're free of cast.

also, if using uint32_t variant, then you can even compare to `UINT32_MAX` in the `if()` below that checks for `offset_in_buffer` being calculated or not.